flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/AdaptedRestartPipelinedRegionStrategyNG.java

```diff
@@ -0,0 +1,302 @@
+		restartTasks(tasksToRestart);
+	}
+
+	private void restartTasks(final Set<ExecutionVertexID> verticesToRestart) {
+		// record current versions
+		final long globalModVersion = executionGraph.getGlobalModVersion();
+		final Set<ExecutionVertexVersion> vertexVersions = new HashSet<>(
+			executionVertexVersioner.recordVertexModifications(verticesToRestart).values());
+
+		// cancel tasks involved in this task failure
+		cancelTasks(verticesToRestart)
+			.thenAccept(
+				(Object ignored) -> {
+					// skip local failover if is in global failover
+					if (!isLocalFailoverValid(globalModVersion)) {
+						LOG.info("Skip current region failover as a global failover is ongoing.");
+						return;
+					}
+
+					// found out vertices which are still valid to restart.
+					// some vertices involved in this failover may be modified if another region
+					// failover happens during the cancellation stage of this failover.
+					// Will ignore the modified vertices as the other failover will deal with them.
+					final Set<ExecutionVertex> unmodifiedVertices = executionVertexVersioner
+						.getUnmodifiedExecutionVertices(vertexVersions)
+						.stream()
+						.map(id -> getExecutionVertex(id))
+						.collect(Collectors.toSet());
+
+					try {
+						LOG.info("Finally restart {} tasks to recover from task failure.", unmodifiedVertices.size());
+
+						// reset tasks to CREATED state and reload state
+						resetTasks(unmodifiedVertices, globalModVersion);
+
+						// re-schedule tasks
+						rescheduleTasks(unmodifiedVertices, globalModVersion);
+					} catch (GlobalModVersionMismatch e) {
+						// happens when a global recovery happens concurrently to the regional recovery
+						// just stop this local failover
+						return;
+					} catch (Exception e) {
+						throw new CompletionException(e);
+					}
+				})
+			.whenComplete(
+				(Object ignored, Throwable t) -> {
+					// fail globally if any error happens
+					if (t != null) {
+						LOG.info("Unexpected error happens in region failover. Fail globally.", t);
+						failGlobal(t);
+					}
+				});
+	}
+
+	@VisibleForTesting
+	protected CompletableFuture<?> cancelTasks(final Set<ExecutionVertexID> verticesToRestart) {
+		final List<CompletableFuture<?>> cancelFutures = verticesToRestart.stream()
+			.map(this::cancelExecutionVertex)
+			.collect(Collectors.toList());
+
+		return FutureUtils.combineAll(cancelFutures);
+	}
+
+	private void resetTasks(
+		final Set<ExecutionVertex> verticesToRestart,
+		final long globalModVersion) throws Exception {
+
+		final Set<CoLocationGroup> colGroups = new HashSet<>();
+		final long restartTimestamp = System.currentTimeMillis();
+
+		for (ExecutionVertex ev : verticesToRestart) {
+			CoLocationGroup cgroup = ev.getJobVertex().getCoLocationGroup();
+			if (cgroup != null && !colGroups.contains(cgroup)){
+				cgroup.resetConstraints();
+				colGroups.add(cgroup);
+			}
+
+			ev.resetForNewExecution(restartTimestamp, globalModVersion);
+		}
```

Comment:
code in line 167~178 is a duplicate of that in `FailoverRegion#reset`, which could be encapsulated as a static function and called by both `AdaptedRestartPipelinedRegionStrategyNG` and `FailoverRegion`.

	static void resetForNewRestartVertices(final Collection<ExecutionVertex> vertices, long globalModVersionOfFailover) throws Exception {
		final Collection<CoLocationGroup> colGroups = new HashSet<>();
		final long restartTimestamp = System.currentTimeMillis();

		for (ExecutionVertex ev : vertices) {
			CoLocationGroup cgroup = ev.getJobVertex().getCoLocationGroup();
			if (cgroup != null && !colGroups.contains(cgroup)){
				cgroup.resetConstraints();
				colGroups.add(cgroup);
			}

			ev.resetForNewExecution(restartTimestamp, globalModVersionOfFailover);
		}
	}

Reply:
This strategy should be a replacement of the legacy RestartPipelinedStrategy, as well as its sub components like FailoverRegion. And it's a temporary solution as well.
In the next few versions, I think we may gradually remove legacy RestartPipelinedStrategy, then remove this adaptor, and then use the new generation FailoverStrategy.

To keep simplicity, I think we'd better not changing existing logic, or as less as possible.